river/IdleInhibitorManager.zig

```diff
@@ -0,0 +1,61 @@
+const Self = @This();
+
+const std = @import("std");
+const wlr = @import("wlroots");
+const wl = @import("wayland").server.wl;
+
+const server = &@import("main.zig").server;
+const util = @import("util.zig");
+
+const IdleInhibitor = @import("IdleInhibitor.zig");
+
+idle_inhibit_manager: *wlr.IdleInhibitManagerV1,
+new_idle_inhibitor: wl.Listener(*wlr.IdleInhibitorV1),
+inhibitors: std.TailQueue(IdleInhibitor) = .{},
+idle: *wlr.Idle,
+
+pub fn init(self: *Self, idle: *wlr.Idle) !void {
+    self.idle_inhibit_manager = try wlr.IdleInhibitManagerV1.create(server.wl_server);
+    self.idle = idle;
+    self.new_idle_inhibitor.setNotify(handleNewIdleInhibitor);
+    self.idle_inhibit_manager.events.new_inhibitor.add(&self.new_idle_inhibitor);
+}
+
+pub fn deinit(self: *Self) void {
+    while (self.inhibitors.pop()) |inhibitor| {
+        util.gpa.destroy(inhibitor);
+    }
+}
+
+pub fn idleInhibitCheckActive(self: *Self) void {
+    var inhibited = false;
+    var it = self.inhibitors.first;
+    while (it) |node| : (it = node.next) {
+        // If for whatever reason the inhibitor does not have a view, then
+        // assume it is visible.
+        if (node.data.view == null) {
+            inhibited = true;
+            break;
+        }
+        // If view is visible,
+        if (node.data.view.?.current.tags & node.data.view.?.output.current.tags != 0) {
+            inhibited = true;
+            break;
+        }
```

Comment:
This check isn't really correct as it doesn't take e.g. layer surfaces or popups into account and doesn't check for visibility of views beyond matching them against the output's focused tags. A view might be hidden behind another opaque view even if it matches the output's tags.

However, there isn't really a way to do this properly and there won't be until we have a proper scene graph. Wlroots has a scene graph API now, but I haven't gotten around to updating river to use it yet.

Reply:
Yea, I wasn't too sure how river was detecting visible views, but that was what I found. Out of curiosity, where can I read more about scene graphs, and is there an open issue for this?

Reply:
Nope, no open issue. It's the next big think I plan to work on for river though.

I pretty much learned about scene graphs by contributing to the wlroots scene graph implementation. I suppose the wikipedia page might be useful for a general overview though: https://en.wikipedia.org/wiki/Scene_graph